### ISSUE TITLE
Tar file not deleted from agent home

### ISSUE THREAD

**Issue by sorin-pop** [djavorszky/ddn#261]
I'm not sure, but I might have stumbled upon a bug while working on #213. I took the ungzip func as template for a new, unbzip2 func. During testing I noticed that when importing a tar.bz2 file, the tar file obtained from the tar.bz2 file is created in the agent home folder (why not under dumps? But actually this is not the real problem, and it doesn't really matter) The problem seems to be that the tar file remains there after the import, it is not deleted.

Then I tested with a tar.gz file (which uses the already existing ungzip func), and I could see the same problem, on my machine at least.

@djavorszky do you think this is indeed a real problem, or for some reason I only see it in my local test environment? Because if it exists in production as well, I guess some of the agents will have collected a few such tar files in their home directories, which would be quite grave, because they take up space unnecessarily. I only have access to the Oracle agent in prod, and indeed, I see there two such tar files: dump6.2 (1).tar and dump6.tar.

Can you please check the other production agents?

**Comment by sorin-pop** [comment]
Looking further at the code, this should be handled by the defer os.Remove(path) line at the beginning of untar func... But why doesn't it happen? Could it be only a Windows issue?...

**Comment by sorin-pop** [comment]
I did some Googling and indeed there seems to be an issue with os.Remove() on Windows. However, Itried various things and couldn't get it to work. @djavorszky can we have a look at this together sometime? It would be important, because it seems that the current code is leaving these files on a Windows agent (and we have Oracle on it, and probably MSSQL will also be on it)

**Comment by djavorszky** [comment]
I remember this coming up with Oracle, that the datafiles weren't deleted. If I remember correctly, we agreed that you'd manually delete them from time to time, as that's easier and less effort than trying to work around it.

Anyway, does this mean that Oracle's home folder has a couple of .tar files? Also, can you share what you've found during your search?

So how do we go forward.. I'm pretty sure we can work this around in Go, but we may need to get a little bit creative. For that, can you share all information you have on the issue? e.g. error messages when trying to remove (for that, you could alter the `defer os.Remove(file)` by doing something like:

```
...
defer doRemove(file)
...

func doRemove(file) {
   err := os.Remove(file)
   if err != nil {
      logger.Error("%#v", err)
   }
}
```

This way, we'd get the error message as well.

**Comment by sorin-pop** [comment]
The other issue was somewhat different. It was about "DROP TABLESPACE INCLUDING CONTENTS AND DATAFILES" not actually deleting the datafiles.

This is about Go's os.Remove(file) in our io.go/untar() func not removing the intermediate tar file extracted from a tar.gz (or tar.bz2). Without any error thrown, apparently. But I will try your suggestion above.

Here is the most relevant link I found about this: https://groups.google.com/forum/#!topic/golang-nuts/9pqdrrDOsNs

**Comment by sorin-pop** [comment]
Also: https://github.com/golang/go/issues/9606

**Comment by sorin-pop** [comment]
I got this error logged with your method above:

&os.PathError{Op:"remove", Path:"./PUSHTEST.tar", Err:0x20}

**Comment by djavorszky** [comment]
Well, according to [Microsoft](https://msdn.microsoft.com/en-us/library/windows/desktop/ms681382(v=vs.85).aspx), 0x20 is:
> ERROR_SHARING_VIOLATION
>   32 (0x20)
>   The process cannot access the file because it is being used by another process.

Right, I figured it out. The problem is that the .tar file is still open when untar tries to delete it, because ungzip still have not closed it. Can you try to close the file before the untar method is called, to see if that works?

**Comment by sorin-pop** [comment]
It seems that this works, thanks Dani! I'll send the fix with the pull request on #213
